**Hello Osama!**

Alexey Kargin is here and I will review your **SQL project** and help you make it even better! Nice to meet you again! I have added my comments on your new project to cells with different coloring. Please, do not move or delete them - it helps us move faster:

<div class="alert alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 I am using green color for my comment, if everything is done successfully. Also, I use this color for general recommendations and additional information.
</div>

<div class="alert alert-warning">
<b>Reviewer's comment v.1</b> 
    
📝 I am using yellow color for my comment if I can give you a recommendation or think that something can be improved in this project. These are optional recommendations, but it will be perfect if you work with them.
</div>

<div class="alert alert-danger">
<b>Reviewer's comment v.1</b> 
    
✍ I am using red color for my comment if the block requires some extra work or corrections. The project couldn’t be accepted with red comments. Also, I highlight some issues with result interpretations using this block too. 
</div>
  
To make a dialog, let’s use this style for your remarks:
<div class="alert alert-info"> 
<b>Student's comment</b> 
    
Just like this. To make a similar block - double-click on the block, copy and edit it in places you are changing. Also, if you have any questions, you could highlight them using this block.
</div>

Now, let’s go to data analysis!

---
    
<div class="alert alert-success">
<b>General feedback v.1</b>
    
👋 Osama, thank you for sending the project! 
    
You have provided a good job! You correctly used subqueries to solve the tasks. <span class="text-danger">However, the second and query should be improved. Please check my comments. </span> and I believe you can easily fix them!  

Keep up the good work, and come back soon! 
    
<br>
Best regards,
<br> Alexey
</div>

---

    
<div class="alert alert-success">
<b>General feedback v.2</b>
    
👋 Osama,  thank you for sending the revised version of the project 👍

Your corrections look great, and now the project is a true "A". Congratulations! I am glad to say that your project has been accepted. You hit the target!


<br>
Best regards,
<br> Alexey
</div>

---

# SQL Project

Table of Contents:

<a href='#Step 1 Describe the Goals of the Study'>Step 1 Describe the Goals of the Study</a>

<a href='#Step 2 Study the Tables'>Step 2 Study the Tables</a>

<a href='#Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook'>Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook</a>

<a href='#Step 4 Conclusions and Recommendations'>Step 4 Conclusions and Recommendations</a>



<a id='Step 1 Describe the Goals of the Study'><a/>

## Project Description

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

I've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 Good start! </div>

In [1]:
# import libraries and establish connection to the database
import psycopg2
import pandas as pd
import numpy as np
# import seaborn as sns
# import matplotlib.pyplot as plt
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})


<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 Well done! The libraries have been loaded, and the connector to the database has been created. Let's study the data!
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment v.1</b> 
    
📝 It is better to aggregate import of libraries into a separate cell at the start of the code. That will be in line with [PEP 8 - Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/). Also, it is not necessary to load libraries that we will not use.
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.2</b> 
    
👍 Thank you for checking this point. </div>

In [3]:
# use this code for SQL queries: pd.io.sql.read_sql(query, con = engine)
# the unction reads a SQL query or database table into a pandas dataframe

<a id='Step 2 Study the Tables'><a/>

## Step 2 Study the Tables

In [4]:
# displaying ERD for the database using markdown

<img src="Data diagram.png">

To begin I'm going to get a look at all the tables in the database.

I'll proceed in this order:
1. Books
2. Authors
3. Publishers
4. Ratings
5. Reviews

In [5]:
#querying the books table
books_query = "SELECT * FROM books"

In [6]:
#creating variable for the book table
books_table = pd.io.sql.read_sql(books_query, con = engine)
books_table.head(5)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [7]:
#querying the authors table
authors_query = "SELECT * FROM authors"

In [8]:
authors_table = pd.io.sql.read_sql(authors_query, con = engine)
authors_table.head(5)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
#querying the publishers table
publishers_query = "SELECT * FROM publishers"

In [10]:
publishers_table = pd.io.sql.read_sql(publishers_query, con = engine)
publishers_table.head(5)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
#querying the publishers table
ratings_query = "SELECT * FROM ratings"

In [12]:
ratings_table = pd.io.sql.read_sql(ratings_query, con = engine)
ratings_table.head(5)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
#querying the publishers table
reviews_query = "SELECT * FROM reviews"

In [14]:
reviews_table = pd.io.sql.read_sql(reviews_query, con = engine)
reviews_table.head(5)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<a id='Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook'><a/>

<a id='Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook'><a/>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment v.1</b> 
    
👍 Well done! The data have been observed. Here we could also provide a data description. Also, when we work with databases using SQL, it is better to limit the data using SQL queries rather than pandas. 
    
```SQL
SELECT *
FROM reviews    
LIMIT 5
```  
    
It allows us to avoid errors and save time if the database has many entries.
    
---
    
    
📝 Also,  we could provide a data description. 
    
</div>

## Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook

Tasks:
1. Find the number of books released after January 1, 2000.
2. Find the number of books released after January 1, 2000.
3. Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
4. Identify the author with the highest average book rating (look only at books with at least 50 ratings).
5. Find the average number of text reviews among users who rated more than 50 books. 

### Task 1: Find the number of books released after January 1, 2000

In [15]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [16]:
#querying the data
books_after_query = "SELECT * FROM books WHERE publication_date > '2000-01-01'"
books_after_2020_01_01 = pd.io.sql.read_sql(books_after_query, con = engine)

In [17]:
print("The number of books released after January 1, 2000 is:", len(books_after_2020_01_01))

The number of books released after January 1, 2000 is: 819


<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 <b>1st query:</b> Correct!
</div>

###  Tasks 2: Find the number of user reviews and the average rating for each book

In [18]:
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [19]:
reviews_table.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [20]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [21]:
book_ratings_query = '''
            SELECT 
                books.title AS title,
                AVG(ratings.rating) as avg_rating,
                COUNT(DISTINCT reviews.review_id) as num_reviews
            FROM books
            INNER JOIN ratings on ratings.book_id = books.book_id
            INNER JOIN reviews on reviews.book_id = books.book_id
            GROUP BY books.book_id
            ORDER BY num_reviews DESC'''

result = pd.io.sql.read_sql(book_ratings_query, con = engine)
display(result.head(10))

title  avg_rating  num_reviews
0                            Twilight (Twilight  #1)    3.662500            7
1                                           The Road    3.772727            6
2                                     The Book Thief    4.264151            6
3                                   The Glass Castle    4.206897            6
4                                Water for Elephants    3.977273            6
5             The Da Vinci Code (Robert Langdon  #2)    3.830508            6
6                             The Catcher in the Rye    3.825581            6
7  The Curious Incident of the Dog in the Night-Time    4.081081            6
8                          Outlander (Outlander  #1)    4.125000            6
9  Harry Potter and the Chamber of Secrets (Harry...    4.287500            6

We can see that top 3 leaders for the reviews amount are following books:<br><br>
- Twilight: 7 reviews with average rating of 3.66;
- The Da Vinci Code: 6 reviews with average rating of 3.83;
- Eat Pray Love: 6 reviews with average rating of 3.39.

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.2</b> 
    
👍 Perfect! Now the query is correct! </div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment v.1</b> 
    
✍ <b>2nd query:</b> I would suggest that this query is not correct. Let's see, on the `review_num` data: there are a lot of reviews. If we calculate the sum of reviews, we will have their total number (`26167`) higher than the number of reviews in the source table (`2793`). Please see my example and fix the query. I guess that it is better to count distinct reviews.

--- 
    
📝 In addition - it is better to format the SQL query within several lines. For that we could use special resources like that https://sqlformat.org/
    
</div>

In [22]:
# the reviewer's code:

# based on the project's query:
display(pd.io.sql.read_sql('''
    WITH count as (
    SELECT books.book_id AS book_id,
       books.title AS title,
       COUNT(reviews.review_id) AS review_count,
       AVG(ratings.rating) AS avg_book_rating
    FROM books
    LEFT JOIN ratings ON ratings.book_id = books.book_id
    LEFT JOIN reviews ON reviews.book_id = books.book_id
    GROUP BY books.book_id
    ORDER BY review_count DESC)
    SELECT SUM(review_count)
    FROM count;
''', con=engine))

# but the source table contains:
display(pd.io.sql.read_sql('''
    SELECT COUNT(reviews)
    FROM reviews;
''', con=engine))

sum
0  26167.0

count
0   2793

### Task 3: Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis)

In [23]:
publishers_table.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [24]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [25]:
publisher_books_query = '''
SELECT publishers.publisher_id AS publisher_id, publishers.publisher AS publisher_name, COUNT(books.book_id) AS book_count 
FROM publishers LEFT JOIN books ON books.publisher_id = publishers.publisher_id 
WHERE books.num_pages > 50 GROUP BY publishers.publisher_id 
ORDER BY book_count DESC LIMIT 1;
'''

In [26]:
publisher_book_count = pd.io.sql.read_sql(publisher_books_query, con = engine)
publisher_book_count.head(10)

publisher_id publisher_name  book_count
0           212  Penguin Books          42

Penguin Books is the publisher with the greatest number of books over 50 pages. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 <b>3rd query:</b> Correct! 
</div>

### Task 4: Identify the author with the highest average book rating (look only at books with at least 50 ratings)

In [27]:
authors_table.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [28]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [29]:
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [30]:
authors_bookratings_query = '''
SELECT books.book_id AS book_id, books.title AS book_title, authors.author_id AS author_id, authors.author AS author_name, AVG(ratings.rating) AS avg_book_rating, COUNT(ratings.rating) AS number_of_ratings 
FROM books LEFT JOIN authors ON authors.author_id = books.author_id LEFT JOIN ratings ON ratings.book_id = books.book_id 
GROUP BY books.book_id, authors.author_id 
HAVING COUNT(ratings.rating_id) >= 50 
ORDER BY avg_book_rating DESC, number_of_ratings DESC LIMIT 5;
'''

In [31]:
author_ratings = pd.io.sql.read_sql(authors_bookratings_query, con = engine)
author_ratings.head(10)

book_id                                         book_title  author_id  \
0      302  Harry Potter and the Prisoner of Azkaban (Harr...        236   
1      722  The Fellowship of the Ring (The Lord of the Ri...        240   
2      299  Harry Potter and the Chamber of Secrets (Harry...        236   
3      656                                     The Book Thief        402   
4      300  Harry Potter and the Half-Blood Prince (Harry ...        236   

                         author_name  avg_book_rating  number_of_ratings  
0         J.K. Rowling/Mary GrandPré         4.414634                 82  
1                     J.R.R. Tolkien         4.391892                 74  
2         J.K. Rowling/Mary GrandPré         4.287500                 80  
3  Markus Zusak/Cao Xuân Việt Khương         4.264151                 53  
4         J.K. Rowling/Mary GrandPré         4.246575                 73

J.K. Rowling has he highest average book rating of 4.414634 for Harry Potter and the Prisoner of Azkaban. This was followed by The Fellowship of the Ring by J.R.R. Tolkien at 4.391892 and then J.K. Rowling again with Harry Potter and the Chamber of Secrets at 4.287500.

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 <b>4th query:</b> Overall, the query is correct. However, please note that this task could have been done differently. Here we calculated the average rating for each book and select the one with higher average rating. But our rating contains several books from a single author and in this case we could aggregate the data for each author, like that:    
</div>

In [32]:
# the reviewer's code:

# based on the project's query:
display(pd.io.sql.read_sql('''
    WITH book_ratings as (
    SELECT books.book_id AS book_id,
       books.title AS book_title,
       authors.author_id AS author_id,
       authors.author AS author_name,
       AVG(ratings.rating) AS avg_book_rating,
       COUNT(ratings.rating) AS number_of_ratings
    FROM books
    LEFT JOIN authors ON authors.author_id = books.author_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
    GROUP BY books.book_id,
             authors.author_id
    HAVING COUNT(ratings.rating_id) >= 50
    ORDER BY avg_book_rating DESC,
             number_of_ratings DESC)
        
    SELECT author_name, 
           AVG(avg_book_rating) AS avg_rating, 
           COUNT(book_title) AS books_count
    FROM book_ratings
    GROUP BY author_name
    ORDER BY avg_rating DESC;   
''', con=engine))

author_name  avg_rating  books_count
0                          J.K. Rowling/Mary GrandPré    4.283844            4
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151            1
2                                      J.R.R. Tolkien    4.258446            2
3                                   Louisa May Alcott    4.192308            1
4                                        Rick Riordan    4.080645            1
5                                     William Golding    3.901408            1
6                                       J.D. Salinger    3.825581            1
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474            1
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879            1
9                                           Dan Brown    3.754540            2
10                                         Lois Lowry    3.750000            1
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730            1
12                                    Stephenie Meyer    3.662500            1
13                                     John Steinbeck    3.622951            1

### Task 5: Find the average number of text reviews among users who rated more than 50 books

In [33]:
reviews_table.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [34]:
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [35]:
average_text_query = '''
SELECT COUNT(*) / COUNT(distinct username) AS average_reviews 
FROM reviews 
WHERE username IN (SELECT username FROM ratings 
GROUP BY username 
HAVING COUNT(*) > 50)
'''

In [36]:
text_reviews_average = pd.io.sql.read_sql(average_text_query, con = engine)
print(text_reviews_average)

   average_reviews
0               24


The average number of text reviews among users who rated 50 or more books is 24.

<a id='Step 4 Conclusions and Recommendations'><a/>

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 <b>5th query:</b> The result is correct! We can also round the result to two decimal places.
    
</div>

## Step 4 Conclusions and Recommendations

In conclusion it is clear that the users of this book review service are quite active, which means that this is an ideal market moving forward. I would recommend the company to give exclusive deals to users who review at least 50 books and use that as an incentive to purchase more books and be more active on the site. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment v.1</b> 
    
👍 Good job!</div>

<a href='#Top of Page'>Top of Page<a/>